src/main/java/graphql/schema/validation/AppliedDirectiveArgumentsAreValid.java

```diff
@@ -32,29 +35,56 @@ public TraversalControl visitGraphQLDirective(GraphQLDirective directive, Traver
         if (context.getParentNode() != null) {
             for (GraphQLArgument graphQLArgument : directive.getArguments()) {
-                checkArgument(directive, graphQLArgument, context);
+                checkArgument(
+                        directive.getName(),
+                        graphQLArgument.getName(),
+                        graphQLArgument.getArgumentValue(),
+                        graphQLArgument.getType(),
+                        context
+                );
             }
         }
         return TraversalControl.CONTINUE;
     }
 
-    private void checkArgument(GraphQLDirective directive, GraphQLArgument argument, TraverserContext<GraphQLSchemaElement> context) {
-        if (!argument.hasSetValue()) {
-            return;
+    @Override
+    public TraversalControl visitGraphQLAppliedDirective(GraphQLAppliedDirective directive, TraverserContext<GraphQLSchemaElement> context) {
+        // if there is no parent it means it is just a directive definition and not an applied directive
+        if (context.getParentNode() != null) {
+            for (GraphQLAppliedDirectiveArgument graphQLArgument : directive.getArguments()) {
+                checkArgument(
+                        directive.getName(),
+                        graphQLArgument.getName(),
+                        graphQLArgument.getArgumentValue(),
+                        graphQLArgument.getType(),
+                        context
+                );
+            }
         }
+        return TraversalControl.CONTINUE;
+    }
+
+    private void checkArgument(
+            String directiveName,
+            String argumentName,
+            InputValueWithState argumentValue,
+            GraphQLInputType argumentType,
+            TraverserContext<GraphQLSchemaElement> context
+    ) {
         GraphQLSchema schema = context.getVarFromParents(GraphQLSchema.class);
         SchemaValidationErrorCollector errorCollector = context.getVarFromParents(SchemaValidationErrorCollector.class);
-        InputValueWithState argumentValue = argument.getArgumentValue();
         boolean invalid = false;
         if (argumentValue.isLiteral() &&
-                !validationUtil.isValidLiteralValue((Value<?>) argumentValue.getValue(), argument.getType(), schema, GraphQLContext.getDefault(), Locale.getDefault())) {
+                !validationUtil.isValidLiteralValue((Value<?>) argumentValue.getValue(), argumentType, schema, GraphQLContext.getDefault(), Locale.getDefault())) {
             invalid = true;
         } else if (argumentValue.isExternal() &&
-                !isValidExternalValue(schema, argumentValue.getValue(), argument.getType(), GraphQLContext.getDefault(), Locale.getDefault())) {
+                !isValidExternalValue(schema, argumentValue.getValue(), argumentType, GraphQLContext.getDefault(), Locale.getDefault())) {
+            invalid = true;
+        } else if (argumentValue.isNotSet() && GraphQLTypeUtil.isNonNull(argumentType))  {
```

Comment:
I suspect we need to do some more work here on how default values get put into place at runtime.  Because if its not null but it has a default than thats ok right.

Reply:
Actually look at the code - we make an assumption that the default value is transferred into the `GraphQLArgument` / `GraphQLAppliedDirectiveArgument` at creation time and hence it should be validated if its a non null type